### DISCUSSION TITLE
Keep getting the error message: Missing or invalid OAuth2 subscription client,

### DISCUSSION THREAD

**Discussion by AramBP** [pocketbase/pocketbase#4589] [Q&A]
Hello,

Im trying to implement oAuth2 in my nextjs app. Ive followed the documentation and am making a call in my code like this:
`try {
      const result = await pb
        .collection("users")
        .authWithOAuth2({ provider: "google" });`
        
with my redirectUrl set like this : 
`  const redirectUrl = "http://127.0.0.1:8090/api/oauth2-redirect";`

But I keep getting this error in pocketbase:
![Screenshot from 2024-03-18 23-49-06](https://github.com/pocketbase/pocketbase/assets/61063343/b5c6a664-eb6f-4130-a4bf-62a93d45c069)

Does anyone know whats going on?

**Comment by ganigeorgiev** [comment]
Without details what your code is doing and where is running (browser, node or both?) it is difficult to pinpoint the exact reason for the above error.

This error is usually returned when the realtime connection has been closed while the OAuth2 signin/signup was still in progress.

Note that the "all-in-one" OAuth2 flow can be used only client-side because it needs to:
- open a browser window with the OAuth2 provider auth page
- establish a realtime connection with the PocketBase server to receive the OAuth2 auth result

If you want to handle the OAuth2 redirect on the server-side, you'll have to use the "Manual code exchange" (the second tab in [OAuth2 integration](https://pocketbase.io/docs/authentication/#oauth2-integration)).

_OAuth2 questions for integrating with "meta" frameworks seems to be common but I really don't know how to explain it better or make it easier. In my opinion a lot of the issues originate from the fact that developers don't realize that when using JS SSR they already have a backend (aka. the node server) or at least not properly weighting the pros and cons when combining with another backend like PocketBase (you could read some of my other ramblings in https://github.com/pocketbase/pocketbase/discussions/4412#discussioncomment-8575834, https://github.com/pocketbase/pocketbase/discussions/4514#discussioncomment-8704533 and other linked discussions)._

**Reply by AramBP** [reply]
Hello thanks for your answer,

Im making my app in nextjs and the function I showed in the discussion is being called on the client side and its running inside my browser. The problem seems to be that pocketbase will establish a real time connection with my app then update it, make a POST request to /api/realtime and then close the realtime connection(message: cancelled request). 
Because of the cancelled request the call fails. 

this my loginOAuth function:
![Screenshot from 2024-03-19 08-16-51](https://github.com/pocketbase/pocketbase/assets/61063343/d4f9225f-59b1-498a-be9e-4ed73bf4abd8)

And im calling it like so:
![Screenshot from 2024-03-19 08-17-36](https://github.com/pocketbase/pocketbase/assets/61063343/b2ae923c-7eb8-4faf-94e1-79d3d4d85b25)

I have autoCancellation set to false like this:
`const pb = new PocketBase("http://127.0.0.1:8090");
 pb.autoCancellation(false);
 `
 I also keep getting this warning in my DevTools console:
 
![Screenshot from 2024-03-19 08-10-29](https://github.com/pocketbase/pocketbase/assets/61063343/45346306-4cf1-434c-952b-a5018f3af9e3)

**Reply by ganigeorgiev** [reply]
I'm not sure how to interpret the above screenshots but I don't think it is JS SDK related.

> The problem seems to be that pocketbase will establish a real time connection with my app then update it, make a POST request to /api/realtime and then close the realtime connection(message: cancelled request).

I don't think this is what is happening. ~Unless you have multiple different client-side SDK instances running at the same time~(_actually even with this it shouldn't be possible to trigger it in this scenario_), the auto cancellation shouldn't be triggered in this case.

If you think that it is a JS SDK issue, feel free to share a **minimal** reproducible example and I'll try to investigate it in more details but without a code sample it is difficult to try to guess what is causing it to fail.

**Reply by ganigeorgiev** [reply]
Also note that `(message: cancelled request)` is the server-side log - it just indicates that the connection between the browser and the PocketBase server was dropped.

If it was autocancellation the cause of the closed connection then `err.isAbort` in your above code would have been `true`.

**Reply by ganigeorgiev** [reply]
Also I just noticed from the screenshot that you have both `a` and `button` handler which is wrong and unnecessary. When using the "all-in-one" flow calling `authWithOAuth2` is all you need.

In any case, the easiest way to rule out PocketBase or JS SDK issue is to create a plain `pb_public/index.html` file with the below content:

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>OAuth2 test</title>
</head>
<body>
    <button id="btn">OAuth2 login</button>

    <script type="module">
        import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@0.21.1/dist/pocketbase.es.mjs"

        const pb = new PocketBase("http://127.0.0.1:8090");

        document.getElementById("btn").addEventListener("click", () => {
            pb.collection("users").authWithOAuth2({
                provider: "google",
            }).then((data) => {
                console.log("Success:", data);
            }).catch((err) => {
                console.log("Failure:", err);
            });
        })
    </script>
</body>
</html>
```

Navigate to `http://127.0.0.1:8090` in your browser, open the inspect console, and click on the "OAuth2 login" button.

**Reply by AramBP** [reply]
Hey man thanks for your help. If anyone is stupid like me make sure that in your google cloud console you have the http://localhost:8090 and http://localhost:3000 urls both set as authorized javascript origins
